CRD Generator -- Stack overflow when using EnumMap type in Spec

### Describe the bug

I'm seeing a stack overflow error in the CrdGenerator when defining a `java.util.EnumMap` as a private field. I believe the cause of this is due to the fact that the library is trying to introspect the `Class<T> keyType` private field that is contained within `EnumMap`. Within the `Class` type is a `ConcurrentHashMap` which has a bunch of circular references and that's where the generator gets into infinite recursion.

This effects multiple versions of the client.

I believe this is caused by part of the design of the CRD generator, where it introspects private fields as the means of generating schemas. The fields that should be added to the spec/status schemas are the ones that will be serialized by Jackson when sending a CR instance to the kube api. The default behavior in Jackson is to look at public, non-static, non-transient fields. This is also the behavior of the serializer that is contained within fabric8. Therefore, the CRD generator should follow the same visibility rules.

I don't think that `private`, `static`, or `transient` properties should be added to the schema, but rather the way schemas are generated should be inverted. First look at public, non-static, non-transient methods (getters) and then check to see if there's a matching private field to pull annotations from. I believe that is the behavior that is expected from jackson.



### Fabric8 Kubernetes Client version

SNAPSHOT

### Steps to reproduce

Add

```
public enum Foo { BAR }

private EnumMap<Foo, String> enumToStringMap;
```

to any Spec or Status. Add it to `io.fabric8.crd.example.map.ContainingMapsSpec` to create a testable reproduction of the issue.



### Expected behavior

1. Stack overflow does not occur
2. Schema generation follows visibility rules of Jackson ObjectMapper

### Runtime

Kubernetes (vanilla)

### Kubernetes API Server version

1.25.3@latest

### Environment

Linux

### Fabric8 Kubernetes Client Logs

_No response_

### Additional context

Stack trace for overflow

```
[ERROR] Tests run: 14, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 5.603 s <<< FAILURE! -- in io.fabric8.crd.generator.ParallelCRDGeneratorTest
[ERROR] io.fabric8.crd.generator.ParallelCRDGeneratorTest.mapPropertyShouldHaveCorrectValueType -- Time elapsed: 2.369 s <<< ERROR!
java.lang.StackOverflowError
        at java.base/java.util.stream.AbstractPipeline.wrapSink(AbstractPipeline.java:522)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at io.sundr.builder.BaseFluent.build(BaseFluent.java:58)
        at io.sundr.model.ClassRefFluentImpl.getArguments(ClassRefFluentImpl.java:227)
        at io.sundr.model.ClassRefBuilder.build(ClassRefBuilder.java:57)
        at io.sundr.model.ClassRefBuilder.build(ClassRefBuilder.java:7)
        at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
        at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
        at io.sundr.builder.BaseFluent.build(BaseFluent.java:58)
        at io.sundr.model.ClassRefFluentImpl.getArguments(ClassRefFluentImpl.java:227)
        at io.sundr.model.ClassRefBuilder.build(ClassRefBuilder.java:57)
        at io.sundr.model.ClassRefBuilder.build(ClassRefBuilder.java:7)
        at io.sundr.model.PropertyFluentImpl.getTypeRef(PropertyFluentImpl.java:259)
...
io.fabric8.crd.generator.visitor.AnnotatedPropertyPathDetector.visitPropertiesClasses(AnnotatedPropertyPathDetector.java:91)
        at io.fabric8.crd.generator.visitor.AnnotatedPropertyPathDetector.visit(AnnotatedPropertyPathDetector.java:76)
        at io.fabric8.crd.generator.visitor.AnnotatedPropertyPathDetector.visit(AnnotatedPropertyPathDetector.java:34)
        at io.sundr.builder.Visitor.visit(Visitor.java:30)
        at io.sundr.builder.VisitorWiretap.visit(VisitorWiretap.java:79)
        at io.sundr.builder.BaseFluent.accept(BaseFluent.java:110)
        at io.sundr.builder.BaseFluent.accept(BaseFluent.java:139)
        at io.sundr.builder.BaseFluent.accept(BaseFluent.java:75)
        at io.fabric8.crd.generator.visitor.AnnotatedPropertyPathDetector.visitPropertiesClasses(AnnotatedPropertyPathDetector.java:91)
        at io.fabric8.crd.generator.visitor.AnnotatedPropertyPathDetector.visit(AnnotatedPropertyPathDetector.java:76)
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,7 @@
 #### Bugs
 
 * Fix #5580: [java-generator] Correctly handle defaults for IntOrString types
+* Fix #5584: Fix CRD generation when EnumMap is used
 
 #### Improvements
 * Fix #5429: moved crd generator annotations to generator-annotations instead of crd-generator-api. Using generator-annotations introduces no transitive dependencies.

--- a/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/AdditionalPrinterColumnDetector.java
+++ b/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/AdditionalPrinterColumnDetector.java
@@ -17,8 +17,6 @@ package io.fabric8.crd.generator.visitor;
 
 import io.fabric8.crd.generator.annotation.PrinterColumn;
 
-import java.util.ArrayList;
-
 public class AdditionalPrinterColumnDetector extends AnnotatedMultiPropertyPathDetector {
 
   public AdditionalPrinterColumnDetector() {
@@ -26,6 +24,6 @@ public class AdditionalPrinterColumnDetector extends AnnotatedMultiPropertyPathD
   }
 
   public AdditionalPrinterColumnDetector(String prefix) {
-    super(prefix, PrinterColumn.class.getSimpleName(), new ArrayList<>());
+    super(prefix, PrinterColumn.class.getSimpleName());
   }
 }

--- a/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/AnnotatedMultiPropertyPathDetector.java
+++ b/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/AnnotatedMultiPropertyPathDetector.java
@@ -22,7 +22,9 @@ import io.sundr.model.Property;
 import io.sundr.model.TypeDef;
 import io.sundr.model.TypeDefBuilder;
 
+import java.util.ArrayDeque;
 import java.util.ArrayList;
+import java.util.Deque;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
@@ -40,21 +42,19 @@ public class AnnotatedMultiPropertyPathDetector extends TypedVisitor<TypeDefBuil
   private final String annotationName;
   private final List<Property> parents;
   private final Map<String, Property> properties;
+  private final Deque<Runnable> toRun;
 
   public AnnotatedMultiPropertyPathDetector(String prefix, String annotationName) {
-    this(prefix, annotationName, new ArrayList<>());
-  }
-
-  public AnnotatedMultiPropertyPathDetector(String prefix, String annotationName, List<Property> parents) {
-    this(prefix, annotationName, parents, new HashMap<>());
+    this(prefix, annotationName, new ArrayList<>(), new HashMap<>(), new ArrayDeque<>());
   }
 
   public AnnotatedMultiPropertyPathDetector(String prefix, String annotationName, List<Property> parents,
-      Map<String, Property> properties) {
+      Map<String, Property> properties, Deque<Runnable> toRun) {
     this.prefix = prefix;
     this.annotationName = annotationName;
     this.parents = parents;
     this.properties = properties;
+    this.toRun = toRun;
   }
 
   private boolean excludePropertyProcessing(Property p) {
@@ -84,15 +84,20 @@ public class AnnotatedMultiPropertyPathDetector extends TypedVisitor<TypeDefBuil
       if (!parents.contains(p) && !excludePropertyProcessing(p)) {
         ClassRef classRef = (ClassRef) p.getTypeRef();
         TypeDef propertyType = Types.typeDefFrom(classRef);
-        if (!propertyType.isEnum()) {
+        if (!propertyType.isEnum() && !classRef.getPackageName().startsWith("java.")) {
           List<Property> newParents = new ArrayList<>(parents);
           newParents.add(p);
-          new TypeDefBuilder(propertyType)
-              .accept(new AnnotatedMultiPropertyPathDetector(prefix, annotationName, newParents, properties))
-              .build();
+          toRun.add(() -> new TypeDefBuilder(propertyType)
+              .accept(new AnnotatedMultiPropertyPathDetector(prefix, annotationName, newParents, properties, toRun)));
         }
       }
     });
+
+    if (parents.isEmpty()) {
+      while (!toRun.isEmpty()) {
+        toRun.pop().run();
+      }
+    }
   }
 
   public Set<String> getPaths() {

--- a/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/AnnotatedPropertyPathDetector.java
+++ b/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/AnnotatedPropertyPathDetector.java
@@ -23,7 +23,9 @@ import io.sundr.model.Property;
 import io.sundr.model.TypeDef;
 import io.sundr.model.TypeDefBuilder;
 
+import java.util.ArrayDeque;
 import java.util.ArrayList;
+import java.util.Deque;
 import java.util.List;
 import java.util.Optional;
 import java.util.concurrent.atomic.AtomicReference;
@@ -39,22 +41,20 @@ public class AnnotatedPropertyPathDetector extends TypedVisitor<TypeDefBuilder> 
   private final String prefix;
   private final String annotationName;
   private final List<Property> parents;
-  private final AtomicReference<Optional<String>> reference;
+  private final AtomicReference<String> reference;
+  private final Deque<Runnable> toRun;
 
   public AnnotatedPropertyPathDetector(String prefix, String annotationName) {
-    this(prefix, annotationName, new ArrayList<>());
-  }
-
-  public AnnotatedPropertyPathDetector(String prefix, String annotationName, List<Property> parents) {
-    this(prefix, annotationName, parents, new AtomicReference<>(Optional.empty()));
+    this(prefix, annotationName, new ArrayList<>(), new AtomicReference<>(), new ArrayDeque<>());
   }
 
   public AnnotatedPropertyPathDetector(String prefix, String annotationName, List<Property> parents,
-      AtomicReference<Optional<String>> reference) {
+      AtomicReference<String> reference, Deque<Runnable> toRun) {
     this.prefix = prefix;
     this.annotationName = annotationName;
     this.parents = parents;
     this.reference = reference;
+    this.toRun = toRun;
   }
 
   private static boolean excludePropertyProcessing(Property p) {
@@ -70,32 +70,10 @@ public class AnnotatedPropertyPathDetector extends TypedVisitor<TypeDefBuilder> 
   public void visit(TypeDefBuilder builder) {
     TypeDef type = builder.build();
     final List<Property> properties = type.getProperties();
-    if (visitProperties(properties)) {
-      return;
-    }
-    visitPropertiesClasses(properties);
+    visitProperties(properties);
   }
 
-  private void visitPropertiesClasses(List<Property> properties) {
-    for (Property p : properties) {
-      if (!(p.getTypeRef() instanceof ClassRef)) {
-        continue;
-      }
-      if (!parents.contains(p) && !excludePropertyProcessing(p)) {
-        ClassRef classRef = (ClassRef) p.getTypeRef();
-        TypeDef propertyType = Types.typeDefFrom(classRef);
-        if (!propertyType.isEnum()) {
-          List<Property> newParents = new ArrayList<>(parents);
-          newParents.add(p);
-          new TypeDefBuilder(propertyType)
-              .accept(new AnnotatedPropertyPathDetector(prefix, annotationName, newParents, reference))
-              .build();
-        }
-      }
-    }
-  }
-
-  private boolean visitProperties(List<Property> properties) {
+  private void visitProperties(List<Property> properties) {
     for (Property p : properties) {
       if (parents.contains(p)) {
         continue;
@@ -107,15 +85,30 @@ public class AnnotatedPropertyPathDetector extends TypedVisitor<TypeDefBuilder> 
         match = annotation.getClassRef().getName().equals(annotationName);
         if (match) {
           newParents.add(p);
-          reference.set(Optional.of(newParents.stream().map(Property::getName).collect(Collectors.joining(DOT, prefix, ""))));
-          return true;
+          reference.set(newParents.stream().map(Property::getName).collect(Collectors.joining(DOT, prefix, "")));
+          return;
+        }
+      }
+
+      if (p.getTypeRef() instanceof ClassRef && !excludePropertyProcessing(p)) {
+        ClassRef classRef = (ClassRef) p.getTypeRef();
+        TypeDef propertyType = Types.typeDefFrom(classRef);
+        if (!propertyType.isEnum() && !classRef.getPackageName().startsWith("java.")) {
+          newParents.add(p);
+          new TypeDefBuilder(propertyType)
+              .accept(new AnnotatedPropertyPathDetector(prefix, annotationName, newParents, reference, toRun));
         }
       }
     }
-    return false;
+
+    if (parents.isEmpty()) {
+      while (!toRun.isEmpty() && reference.get() == null) {
+        toRun.pop().run();
+      }
+    }
   }
 
   public Optional<String> getPath() {
-    return reference.get();
+    return Optional.ofNullable(reference.get());
   }
 }

--- a/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/LabelSelectorPathDetector.java
+++ b/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/LabelSelectorPathDetector.java
@@ -17,8 +17,6 @@ package io.fabric8.crd.generator.visitor;
 
 import io.fabric8.kubernetes.model.annotation.LabelSelector;
 
-import java.util.ArrayList;
-
 public class LabelSelectorPathDetector extends AnnotatedPropertyPathDetector {
 
   public LabelSelectorPathDetector() {
@@ -26,6 +24,6 @@ public class LabelSelectorPathDetector extends AnnotatedPropertyPathDetector {
   }
 
   public LabelSelectorPathDetector(String prefix) {
-    super(prefix, LabelSelector.class.getSimpleName(), new ArrayList<>());
+    super(prefix, LabelSelector.class.getSimpleName());
   }
 }

--- a/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/SpecReplicasPathDetector.java
+++ b/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/SpecReplicasPathDetector.java
@@ -17,8 +17,6 @@ package io.fabric8.crd.generator.visitor;
 
 import io.fabric8.kubernetes.model.annotation.SpecReplicas;
 
-import java.util.ArrayList;
-
 public class SpecReplicasPathDetector extends AnnotatedPropertyPathDetector {
 
   public SpecReplicasPathDetector() {
@@ -26,6 +24,6 @@ public class SpecReplicasPathDetector extends AnnotatedPropertyPathDetector {
   }
 
   public SpecReplicasPathDetector(String prefix) {
-    super(prefix, SpecReplicas.class.getSimpleName(), new ArrayList<>());
+    super(prefix, SpecReplicas.class.getSimpleName());
   }
 }

--- a/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/StatusReplicasPathDetector.java
+++ b/crd-generator/api/src/main/java/io/fabric8/crd/generator/visitor/StatusReplicasPathDetector.java
@@ -17,12 +17,10 @@ package io.fabric8.crd.generator.visitor;
 
 import io.fabric8.kubernetes.model.annotation.StatusReplicas;
 
-import java.util.ArrayList;
-
 public class StatusReplicasPathDetector extends AnnotatedPropertyPathDetector {
 
   public StatusReplicasPathDetector(String prefix) {
-    super(prefix, StatusReplicas.class.getSimpleName(), new ArrayList<>());
+    super(prefix, StatusReplicas.class.getSimpleName());
 
   }
 

--- a/crd-generator/api/src/test/java/io/fabric8/crd/example/map/ContainingMaps.java
+++ b/crd-generator/api/src/test/java/io/fabric8/crd/example/map/ContainingMaps.java
@@ -19,8 +19,16 @@ import io.fabric8.kubernetes.client.CustomResource;
 import io.fabric8.kubernetes.model.annotation.Group;
 import io.fabric8.kubernetes.model.annotation.Version;
 
+import java.util.EnumMap;
+
 @Group("map.fabric8.io")
 @Version("v1alpha1")
 public class ContainingMaps extends CustomResource<ContainingMapsSpec, Void> {
+
+  public enum Foo {
+    BAR
+  }
+
+  private EnumMap<Foo, String> enumToStringMap;
 
 }

